language/move-prover/bytecode-to-boogie/src/bytecode_instrs.bpl

```diff
@@ -15,16 +18,76 @@ function {:constructor} Path(p: [int]Edge, size: int): Path;
 
 const DefaultPath: [int]Edge;
 
+type {:datatype} TypeValue;
+type {:datatype} TypeValueArray;
+function {:constructor} BooleanType() : TypeValue;
+function {:constructor} IntegerType() : TypeValue;
+function {:constructor} AddressType() : TypeValue;
+function {:constructor} ByteArrayType() : TypeValue;
+function {:constructor} StrType() : TypeValue;
+function {:constructor} ReferenceType(t: TypeValue) : TypeValue;
+function {:constructor} VectorType(t: TypeValue) : TypeValue;
+function {:constructor} StructType(name: TypeName, ts: TypeValueArray) : TypeValue;
+function {:constructor} TypeValueArray(v: [int]TypeValue, l: int): TypeValueArray;
+
+const EmptyTypeValueArray: TypeValueArray;
+axiom l#TypeValueArray(EmptyTypeValueArray) == 0;
+axiom (forall v1, v2: [int]TypeValue, l1, l2: int ::
+         TypeValueArray(v1, l1) == TypeValueArray(v2, l2) <==>
+         l1 == l2 && (forall i: int :: i >= 0 && i < l1 ==> v1[i] == v2[i]));
+
 type {:datatype} Value;
+type {:datatype} ValueArray;
 function {:constructor} Boolean(b: bool): Value;
 function {:constructor} Integer(i: int): Value;
 function {:constructor} Address(a: Address): Value;
 function {:constructor} ByteArray(b: ByteArray): Value;
 function {:constructor} Str(a: String): Value;
-function {:constructor} Map(m: [Edge]Value): Value;
-function {:constructor} Vector(v: [Edge]Value, l: int): Value;
-
-const DefaultMap: [Edge]Value;
+function {:constructor} Vector(v: ValueArray): Value;
+function {:constructor} Struct(v: ValueArray): Value;
+function {:constructor} ValueArray(v: [int]Value, l: int): ValueArray;
+
+function vlen(v: Value): int { l#ValueArray(v#Vector(v)) }
```

Comment:
Please add {:inline} attribute on all of these helper functions so Boogie knows to inline them instead of introducing an axiom.

Reply:
Done